Gitlab CI migration

mliszcz opened the pull request:
This PR adds configuration for Gitlab CI/CD pipelines and introduces necessary test code changes to make them pass on Gitlab.

mliszcz committed:
Correct AddLoggingTarget test (cxx_dserver_cmd)

The test must not depend on current user permissions for accessing the
log path.

mliszcz committed:
Add CMake flag for generating code coverage data

mliszcz committed:
Temporarily disable tests failing in Gitlab CI/CD

These tests show timing-related failures on single-core runners provided
by gitlab.com. Further work is needed to fix these tests.

mliszcz committed:
Allow collecting code coverage data in CI

mliszcz committed:
Adapt tests to single-core machines (Gitlab CI/CD)

Some tests exhibit different behavior on a single-core machines in
scenario where a call to poll_attribute is immediately followed by
a call to subscribe_event (for e.g. periodic events).

Most likely the issue is related to the polling algorithm and to the
detection of late events, but further analysis is required to confirm
that.

mliszcz committed:
Add Gitlab CI/CD configuration file

mliszcz committed:
Change install target used in CI

We now install the library using the install/fast target which does not
force us to build the default (all) target. This is useful for debugging
issues in CI.

mliszcz committed:
Document support for test coverage analysis

mliszcz commented:
Note that for now I disabled two tests (one testcase in CXX::cxx_dserver::misc and event::change_event). I do not know how to correct them.

mliszcz commented:
Also note that the same branch is pushed to the test gitlab repository: https://gitlab.com/tango-controls/test_gh_import/cppTango/-/tree/gitlab-ci-migration

mliszcz commented:
And FYI we've just hit the dockerhub pull limit:
```
toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit

The command "docker build --build-arg APP_UID=$(id -u) --build-arg APP_GID=$(id -g) -t cpp_tango .travis/${OS_TYPE}
```

andygotz commented:
Hi @mliszcz - HNY!

How to solve this? Should we upgrade to a TEAM licence or should we host our own docker images? I see the limit is 200 / 6 hours. How many pull requests do we have per build? We could host the docker images on gitlab in the future: https://docs.gitlab.com/ee/user/packages/container_registry/

mliszcz commented:
Hi Andy!

I think for now we don't need to do anything. As you've said, we are just blocked for six hours. I don't know how many pulls we need. @t-b calculated that some time ago and if I recall we were below the current limits. I exhausted this limit only due to extensive use of CI for testing the new Gitlab CI pipeline for cppTango.

Yes, we can move the images to Gitlab container registry if necessary, this is a good idea IMO. We can also keep the images in dockerhub and just mirror them (manually by `docker push` for now) in gitlab so that cppTango CI pipeline will not contribute to the dockerhub pull limits. This would be very simple to do.

In the long run I think we should stop using docker or any containers for tests in CI because it contributes to the execution time and it increases complexity of our setup. But this would require some changes on database ds side (mainly adding support for SQLite).

bourtemb commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,203 @@
+        -e COVERALLS_REPO_TOKEN=${COVERALLS_REPO_TOKEN}
+        -e SONAR_TOKEN=${SONAR_TOKEN}
+        -e TANGO_HOST=${TANGO_HOST}
+        -e TRAVIS_BRANCH=${TRAVIS_BRANCH}
+        -e TRAVIS_JOB_ID=${TRAVIS_JOB_ID}
```
Who is setting these TRAVIS_BRANCH and TRAVIS_JOB_ID variables when running on gitlab?
They seem to be used by sonar.sh.
I guess this will have to be changed for sonar to work properly.
@mliszcz I remember you mentioned on slack in December that we should probably enable Sonar after the migration so we could fix sonar in a PR after the migration (or use other tools to do the same things as sonar).
Maybe we should actually forget about Sonar after the migration and concentrate on issues like https://github.com/tango-controls/cppTango/issues/805

mliszcz replied:
Thanks for pointing those! I haven't noticed these variables when I copy-pasted the instructions. Since it is related to sonar, I propose to leave it as it is, so when we get back to sonar integration after migration we will be able to notice that something still needs to be changed here.

I'm all in for replacing sonar with something that we can run locally and that can produce html report similar to what we have in sonar. It is never good to depend on third-party, proprietary service.

bourtemb commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,203 @@
+    - docker stop cpp_tango
+  artifacts:
+    when: always
+    expire_in: 1h
```
isn't 1h a bit too short?

mliszcz replied:
Right. What do you propose? Maybe 1 day would be ok? I'd rather not abuse GL-provided storage with old test logs and core dumps. Also note that artifacts from the last run of each job are kept indefinitely, regardless of this setting.

bourtemb replied:
I would try not to abuse with core files, which can be big indeed.
But I think having several days would be great (so we can launch something on Friday evening and come back on Monday or Tuesday and get the results).
I'm not sure to understand fully your sentence "Also note that artifacts from the last run of each job are kept indefinitely, regardless of this setting."
What happens if someone makes a new commit in another branch? Will this one be considered as the last run?
Or is it the artifacts from the last run from each branch which are kept?

t-b replied:
Why not just use the default for now? We need to buy more project space anyway soon.

mliszcz replied:
I did as Thomas suggested and removed explicit setting of expiration time.

> What happens if someone makes a new commit in another branch? Will this one be considered as the last run?
Or is it the artifacts from the last run from each branch which are kept?

Per my understanding, it keeps all artifacts from all jobs in the last pipeline of each branch, MR and tag:
https://docs.gitlab.com/ee/ci/yaml/#artifactsexpire_in
https://gitlab.com/gitlab-org/gitlab/-/issues/16267
https://gitlab.com/gitlab-org/gitlab/-/issues/229936

bourtemb commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,203 @@
+    - gzip -r build/cpp_test_suite/test_results
+  artifacts:
+    when: always
+    expire_in: 1h
```
isn't 1h a bit too short?

mliszcz replied:
Removed

bourtemb commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,203 @@
+    - gzip -r compat_reports
+  artifacts:
+    when: always
+    expire_in: 1h
```
isn't 1h a bit too short?
Default is 30 days if we don't precise this keyword I think, which seems more reasonable to me.

mliszcz replied:
Removed

bourtemb reviewed (commented):
Thanks @mliszcz for this great and useful work.
I would suggest to remove the expiration delay of 1h for some artifacts (default seems to be 30 days which seems more convenient, unless I missed 1 constraint).
I would also suggest to create a Merge Request on our Gitlab Test repository to check whether the abi-compliance-check is correctly executed in this case.
By the way, it's great to have converted abi-compliance-check too.
The modifications for code coverage in CMake files will be useful too.

bourtemb commented:
I created the Merge Request on the Gitlab test repo to check whether the tests triggered only on Merge Request are working properly.
Here is the MR: https://gitlab.com/tango-controls/test_gh_import/cppTango/-/merge_requests/819

mliszcz commented:
Thanks for triggering the ABI/API check. It failed:
```
$ docker exec -w /home/tango/src -e CI_TARGET_BRANCH=${CI_MERGE_REQUEST_TARGET_BRANCH_NAME} -e CMAKE_BUILD_PARALLEL_LEVEL=$(nproc) cpp_tango .travis/check-ABI-API-compliance.sh || true
fatal: No such ref: 'origin/tango-9-lts'
```
Either gitlab uses different remote name or does not fetch all branches. I'll have a look right now.

mliszcz committed:
Compress all core file artifacts

mliszcz committed:
Keep CI artifacts for the default time period

mliszcz committed:
Fix API/API compilance check CI job definition

Fetch target branch from the remote. This is required by the script.

mliszcz committed:
Correct rules for including jobs into pipelines

With only/except syntax we would need to specify that for all jobs.
With rules syntax we need some fallback rule to prevent duplicated
pipelines from being generated by the same event (e.g. a push to a
branch for which a MR is opened).

mliszcz commented:
API/API compliance check is fixed now: https://gitlab.com/tango-controls/test_gh_import/cppTango/-/jobs/956206341
By default Gitlab runner is not fetching any branches except the current one.

bourtemb commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,211 @@
+        .travis/check-ABI-API-compliance.sh || true
+  after_script:
+    - docker stop cpp_tango
+    - gzip -r compat_reports
```
If we don't gzip the compat reports, it might be possible to browse them more easily online?: https://gitlab.com/tango-controls/test_gh_import/cppTango/-/jobs/956206341/artifacts/browse/compat_reports/libtango/fe934f0ccf09980e679e1f87b0bd042573a31a9b_to_46b5052e3f4576e0f5274c5d6cbbba58bde6a897/
What do you think?

mliszcz replied:
Yes, compression is not needed here. I guess Gitlab could display file contents directly in the artifacts UI: https://gitlab.com/gitlab-org/gitlab/-/issues/211517

I just pushed a commit that removes compression. Let's wait and see how it looks like: https://gitlab.com/tango-controls/test_gh_import/cppTango/-/jobs/960823927

mliszcz replied:
We can indeed browse html reports directly in the gitlab UI. Thanks for this comment.

https://tango-controls.gitlab.io/-/test_gh_import/cppTango/-/jobs/960823927/artifacts/compat_reports/libtango/fe934f0ccf09980e679e1f87b0bd042573a31a9b_to_a0e273c48570abc5c7820c3ca3dc4b05c3594e29/compat_report.html

mliszcz committed:
Do not compress ABI/API compatibility check reports

These reports are expected to be small and having them uncompressed
could improve browsing experience.

bourtemb commented on INSTALL.md:
```diff
@@ -71,6 +71,7 @@ PKG_CONFIG_PATH="/usr/local/libzmq:/usr/local/omniORB" cmake ..
 | `CPPZMQ_BASE`                |                                        | cppzmq installed path
 | `IDL_BASE`                   |                                        | tangoidl installed path
 | `OMNI_BASE`                  |                                        | omniORB4 installed path
+| `TANGO_ENABLE_COVERAGE`      | `OFF`                                  | Instrument code for coverage analysis (Requres CMake 3.13+)
```
```suggestion
| `TANGO_ENABLE_COVERAGE`      | `OFF`                                  | Instrument code for coverage analysis (Requires CMake 3.13+)
```

bourtemb commented on INSTALL.md:
````diff
@@ -283,3 +284,26 @@ cmake -G "Visual Studio 15 2017 Win64" -DCMAKE_INSTALL_PREFIX=install -DCMAKE_BU
+   in many formats. Note that usually one is interested only in the library
+   code and may want to exclude any test code from the report.
+   ```
+   gcovr --filter '^cppapi/' --filter '^log4tango/' --html-details --output coverage.html
````
We should probably suggest to exclude log4tango/tests directory too in the coverage report.

mliszcz replied:
Done in 13fbbd3e04a0604763866b2ac36394744d63923d

mliszcz committed:
Fix typo in TANGO_ENABLE_COVERAGE description

Co-authored-by: Reynald Bourtembourg <reynald.bourtembourg@esrf.fr>

mliszcz committed:
Exclude log4tango/tests from code coverage reports

t-b reviewed (approved):
Looks good!

I think we should the cleanup (ditching travis in paths and names, remove coveralls/bintray?/sonar? support) once we have migrated to gitlab.

mliszcz commented:
Thanks @t-b! For cleanup tasks I created an issue: https://github.com/tango-controls/cppTango/issues/821
Please use that one to report any things we should remember about after the migration.

bourtemb reviewed (approved):
Looks good to me too!
Thanks @mliszcz

mliszcz commented:
Thank both of you for the review. Today or tomorrow I'm going to port these changes to 9.3-backports and open a similar PR.

mliszcz commented:
@t-b, @bourtemb should I merge that one? Or after the migration? Note that it is disabling two tests.

bourtemb commented:
@mliszcz
For me, it's fine to merge it now.
